warp/yul/ExecEnv.py

```diff
@@ -0,0 +1,233 @@
+        self.last_function: Optional[ast.FunctionDefinition] = None
+        self.curr_func: str = []
+        self.callgraph = {}
+        self.needs_exec_env: list[str] = []
+
+    def traverse_inner(self, calledFuncs):
+        if calledFuncs == []:
+            return
+        else:
+            for func in calledFuncs:
+                if self.callgraph[func]["calldataOp"]:
+                    self.needs_exec_env.append(func)
+                self.traverse_inner(self.callgraph[func]["calledFunctions"])
+
+    def traverse_callgraph(self, callgraph):
+        for k, v in callgraph.items():
+            if k == "fun_ENTRY_POINT":
+                for func in v["calledFunctions"]:
+                    try:
+                        self.needs_exec_env.append(func)
+                        self.traverse_inner(self.callgraph[func]["calledFunctions"])
+                    except KeyError:
+                        continue
+            else:
+                continue
+
+    def get_exec_env_functions(self, node: ast.Node) -> list[str]:
+        self.print(node)
+        self.traverse_callgraph(self.callgraph)
+        return self.needs_exec_env
+
+    def print(self, node: ast.Node, *args, **kwargs) -> str:
+        return self.visit(node, *args, **kwargs)
+
+    def common_visit(self, node: ast.Node, *args, **kwargs):
+        raise AssertionError(
+            f"Each node type should have a custom visit, but {type(node)} doesn't"
+        )
+
+    def visit_typed_name(self, node: ast.TypedName, split: bool = False) -> str:
+        if not split:
+            return f"{node.name} : {node.type}"
+        assert node.type == "Uint256", "Can't split non Uin256 type"
+        # could have added ": felt", but when a type is omitted, it's felt by default
+        return f"{node.name}_low, {node.name}_high"
+
+    def visit_literal(self, node: ast.Literal) -> str:
+        v = int(node.value)  # to convert bools: True -> 1, False -> 0
+        high, low = divmod(v, UINT128_BOUND)
+        return f"Uint256(low={low}, high={high})"
+
+    def visit_identifier(self, node: ast.Identifier) -> str:
+        return f"{node.name}"
+
+    def visit_assignment(self, node: ast.Assignment) -> str:
+        return self.visit(
+            ast.VariableDeclaration(
+                variables=[ast.TypedName(x.name) for x in node.variable_names],
+                value=node.value,
+            )
+        )
+
+    def visit_function_call(self, node: ast.FunctionCall) -> str:
+        fun_repr = self.print(node.function_name)
+        args_repr = ", ".join(self.print(x) for x in node.arguments)
+        if fun_repr == "revert":
+            return "assert 0 = 1\njmp rel 0"
+        if fun_repr == "revert":
+            return ""
+        if "return" in fun_repr:
+            return ""
+        if fun_repr == "checked_add_uint256":
+            fun_repr = "add"
+        if "callvalue" in fun_repr:
+            return "Uint256(0,0)"
+        result: str
```

Comment:
It seems like a lot of string manipulation which is not needed, we never use string output of this class

Reply:
we use it for all the callvalue checks for non-payable functions

Reply:
Oh I see. Can't we just though have a flag `has_calldata_call` and set it when there is a function call that is "calldataload" or "calldatasize"? Set it to false at the beginning of each function. Then query at the end. It seems like it would eliminate duplication between this and ToCairoVisitor. It's your call though.

Reply:
But we would still need to traverse the callgraph to see if a function needs exec_env, because it might call a function which calls a function that needs exec_env.

Reply:
All functions that need calldata also get marked as calldata. It's the same as in RevertNormalizer, the order is guaranteed to be optimal, because the's been top-sorted